packages/kit/src/runtime/server/respond.js

```diff
@@ -59,6 +59,74 @@ const allowed_page_methods = new Set(['GET', 'HEAD', 'OPTIONS']);
  * @returns {Promise<Response>}
  */
 export async function respond(request, options, manifest, state) {
+	return handle_request(request, options, manifest, state);
+}
+
+/**
+ * @param {import('types').SSROptions} options
+ * @param {import('@sveltejs/kit').SSRManifest} manifest
+ * @param {import('types').SSRState} state
+ * @returns {(info: RequestInit | import('crossws').Peer) => Promise<Partial<import('crossws').Hooks> & { upgrade: import('crossws').Hooks['upgrade'] }>}
+ */
+export function get_websocket_hooks_resolver(options, manifest, state) {
+	return async (info) => {
+		/** @type {Request} */
+		let request;
+
+		// Check if info is a Peer object
+		if ('request' in info) {
+			// @ts-ignore UpgradeRequest and Request are essentially the same
+			request = info.request;
+		} else {
+			// @ts-ignore although the type is RequestInit, it is almost always a Request object
+			request = info;
+		}
+
+		const result = await handle_request(request, options, manifest, state, true);
+
+		if (result instanceof Response) {
+			// if the result is a Response instead of WebSocket hooks, it means
+			// we should ignore the upgrade request and send back a regular response
+			return {
+				upgrade: () => {
+					// we have to throw the Response to avoid accepting the upgrade
+					throw result;
+				}
+			};
+		}
+
+		return result;
+	};
+}
+
+// we need the type overload so that TypeScript knows the return value
+// can only be a Response if the upgrade param was omitted
+/**
+ * @overload
+ * @param {Request} request
+ * @param {import('types').SSROptions} options
+ * @param {import('@sveltejs/kit').SSRManifest} manifest
+ * @param {import('types').SSRState} state
+ * @returns {Promise<Response>}
+ */
+/**
+ * @overload
+ * @param {Request} request
+ * @param {import('types').SSROptions} options
+ * @param {import('@sveltejs/kit').SSRManifest} manifest
+ * @param {import('types').SSRState} state
+ * @param {boolean} upgrade
+ * @returns {Promise<Response | import('crossws').Hooks>}
+ */
+/**
+ * @param {Request} request
+ * @param {import('types').SSROptions} options
+ * @param {import('@sveltejs/kit').SSRManifest} manifest
+ * @param {import('types').SSRState} state
+ * @param {boolean=} upgrade
+ * @returns {Promise<Response | import('crossws').Hooks>}
+ */
```

Comment:
These two have the same return type, so should be able to be combined I think?
```suggestion
/**
 * @param {Request} request
 * @param {import('types').SSROptions} options
 * @param {import('@sveltejs/kit').SSRManifest} manifest
 * @param {import('types').SSRState} state
 * @param {boolean} [upgrade]
 * @returns {Promise<Response | import('crossws').Hooks>}
 */
```

Reply:
Unfortunately, we do need both. One is the overload signature (dictates how we can call the function) while the other is the implementation signature (dictates the overloads we can write)
https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

Reply:
ah, I see. thanks for clarifying